src/components/input/index.jsx

```diff
@@ -0,0 +1,171 @@
+import { useEffect, useState, useRef } from "react";
+
+import PropTypes from "prop-types";
+
+const COMPONENT_CLASS_NAME = "c-input";
+
+export const FIELD_TYPES = {
+	EMAIL: "email",
+	PASSWORD: "password",
+	TEXT: "text",
+};
+
+const INPUT_STATUS = {
+	DEFAULT: "default",
+	ERROR: "error",
+	SUCCESS: "success",
+	WARNING: "warning",
+};
+
+const getDerivedInputState = ({ valid, inputState }) => {
+	// regex input validation overrides props input state
+	if (!valid) {
+		return INPUT_STATUS.ERROR;
+	}
+
+	// falls back to default if no props input state
+	return inputState || INPUT_STATUS.DEFAULT;
+};
+
+const Input = ({
+	autoComplete,
+	className,
+	defaultValue,
+	hidden,
+	label,
+	name,
+	onChange,
+	placeholder,
+	required,
+	showDefaultError,
+	tip,
+	type,
+	validationErrorMessage,
+	validationPattern,
+	inputState,
+}) => {
+	const [valid, setValid] = useState(true);
+	const [value, setValue] = useState(defaultValue);
+	const [initialBlur, setInitialBlur] = useState(showDefaultError);
+	const inputElement = useRef();
+
+	useEffect(() => {
+		if (initialBlur) {
+			if (validationErrorMessage) {
+				inputElement.current.setCustomValidity("");
+			}
+			const isValid = inputElement.current?.checkValidity();
+			if (!isValid && validationErrorMessage) {
+				inputElement.current.setCustomValidity(validationErrorMessage);
+			}
+			setValid(isValid);
+		}
+	}, [initialBlur, inputElement, validationErrorMessage, validationPattern, value]);
+
+	const handleBlur = () => {
+		setInitialBlur(true);
+	};
+
+	const handleChange = (event) => {
+		setValue(event.target.value);
+		onChange({ value: event.target.value });
+	};
+
+	const infoId = `id_${name}_error`;
+	const inputId = `id_${name}_input`;
+
+	const fieldParameters = {
+		...(autoComplete ? { autoComplete } : {}),
+		...(defaultValue !== "" ? { defaultValue } : {}),
+		...(validationPattern ? { pattern: validationPattern } : {}),
+		...(placeholder ? { placeholder, "aria-placeholder": placeholder } : {}),
+		...(required ? { required } : {}),
+		...(tip || !valid ? { "aria-describedby": infoId } : {}),
+		...(!valid ? { "aria-invalid": true } : {}),
+	};
+
+	// derive input state for determining variant class names
+	const derivedInputState = getDerivedInputState({ valid, inputState });
+
+	// todo: implement hidden in parent div rather than label, input, etc
+	const containerClassNames = [
+		COMPONENT_CLASS_NAME,
+		className,
+		hidden && `${COMPONENT_CLASS_NAME}--hidden`,
+	]
+		.filter((classString) => classString)
+		.join(" ");
+	return (
+		<div className={containerClassNames}>
+			<label
+				className={`${COMPONENT_CLASS_NAME}__label ${COMPONENT_CLASS_NAME}__label--${derivedInputState}`}
```

Comment:
I am curious if we approach this is a different way - and apply the `${derivedInputState}` class to the parent element rather than each child and use the cascade to it's full advantage. That way if you have 

```
.c-input--error {
  color: red;
}
```

which is then applied to all children elements

Reply:
oh ok. this assumes that your label color would want to be used the same as your tip or error color, right?

Reply:
Right - and we could still add hooks for the label - just thinking through the way an editor would have to configure the least amount of settings.
Setting red color at the parent then having to only change the label if they want that to be different would be less configuration - but also remember we still want the individual control of each item to provide the most flexibility

Reply:
so we want both the parent variations and the children ones?

Reply:
Not exactly - something like this could work 

```css
@use "../../scss";

.c-input {
	@include scss.component-properties("input");

	.c-input__input {
		@include scss.component-properties("input-input");
	}

	.c-input__label {
		@include scss.component-properties("input-label");
	}

	.c-input__tip {
		@include scss.component-properties("input-tip");
	}
}

.c-input--error {
	@include scss.component-properties("input-error");

	.c-input__input {
		@include scss.component-properties("input-error-input");
	}

	.c-input__label {
		@include scss.component-properties("input-error-label");
	}

	.c-input__tip {
		@include scss.component-properties("input-error-tip");
	}
}
```

Reply:
huh. do you think I should go down this for this starting ticket? To achieve the mocks, the implementation that just was pushed (modifying all the children) gets there

Reply:
huh. for achieving the mocks, I guess the parent styling all of the colors wouldn't work though. The label does not become red, for instance, on error. should I iterate through these based on your example?

Reply:
updated @matthewroach